File: e2e/comparators.py
"""Semantic comparators for e2e output comparison.

The harness compares the post-apply working tree against an ``expected/``
tree using a format-aware comparator selected by file extension. See the
e2e-harness spec for the full contract.
"""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

import yaml
from lxml import html as lxml_html
from lxml.etree import _Element


def _read_text(p: Path) -> str:
    return p.read_text().replace("\r\n", "\n")


def compare_text(expected: Path, actual: Path) -> None:
    assert _read_text(expected) == _read_text(actual), (
        f"text content differs: {expected} vs {actual}"
    )


def compare_json(expected: Path, actual: Path) -> None:
    exp = json.loads(expected.read_text())
    act = json.loads(actual.read_text())
    assert exp == act, f"JSON content differs: {expected} vs {actual}"


def compare_yaml(expected: Path, actual: Path) -> None:
    exp = yaml.safe_load(expected.read_text())
    act = yaml.safe_load(actual.read_text())
    assert exp == act, f"YAML content differs: {expected} vs {actual}"


def _normalize_whitespace(text: str | None) -> str:
    if text is None:
        return ""
    return re.sub(r"\s+", " ", text).strip()


def _element_signature(el: _Element) -> dict[str, Any]:
    return {
        "tag": el.tag,
        "attrib": dict(el.attrib),
        "text": _normalize_whitespace(el.text),
        "tail": _normalize_whitespace(el.tail),
        "children": [_element_signature(c) for c in el],
    }


def compare_html(expected: Path, actual: Path) -> None:
    exp_tree = lxml_html.fromstring(expected.read_text())
    act_tree = lxml_html.fromstring(actual.read_text())
    exp_sig = _element_signature(exp_tree)
    act_sig = _element_signature(act_tree)
    assert exp_sig == act_sig, (
        f"HTML structure differs: {expected} vs {actual}\n"
        f"expected={exp_sig!r}\nactual  ={act_sig!r}"
    )


_COMPARATORS = {
    ".json": compare_json,
    ".yaml": compare_yaml,
    ".yml": compare_yaml,
    ".html": compare_html,
    ".htm": compare_html,
}


def compare_tree(expected_dir: Path, actual_dir: Path) -> None:
    """Compare every file under expected_dir against actual_dir."""
    expected_files = sorted(
        p for p in expected_dir.rglob("*") if p.is_file()
    )
    for exp in expected_files:
        rel = exp.relative_to(expected_dir)
        act = actual_dir / rel
        assert act.exists(), f"Missing expected file: {rel} under {actual_dir}"
        comparator = _COMPARATORS.get(exp.suffix.lower(), compare_text)
        comparator(exp, act)

File: tests/e2e/comparators.ts
import { existsSync, readFileSync, readdirSync } from "node:fs";
import path from "node:path";

import { JSDOM } from "jsdom";
import { load } from "js-yaml";
import { expect } from "vitest";

/**
 * Semantic comparators for e2e output comparison.
 *
 * A fixture's `expected/` tree describes the *meaning* of a correct apply, not
 * a byte-for-byte rendering. Key order in JSON, quoting style in YAML and
 * attribute order in HTML are all serialiser details engraft is free to change,
 * so each format gets a comparator that looks past them. Anything without a
 * known format falls back to a text comparison with line endings normalised.
 */

const readText = (file: string): string => readFileSync(file, "utf8").replaceAll("\r\n", "\n");

const compareText = (expected: string, actual: string): void => {
  expect(readText(actual), `text content differs: ${actual}`).toBe(readText(expected));
};

const compareJson = (expected: string, actual: string): void => {
  expect(JSON.parse(readText(actual)), `JSON content differs: ${actual}`).toEqual(
    JSON.parse(readText(expected)),
  );
};

const compareYaml = (expected: string, actual: string): void => {
  expect(load(readText(actual)), `YAML content differs: ${actual}`).toEqual(
    load(readText(expected)),
  );
};

const normalizeWhitespace = (text: string | null): string =>
  (text ?? "").replace(/\s+/gu, " ").trim();

type HtmlNode = string | { comment: string } | HtmlElementSignature;

interface HtmlElementSignature {
  tag: string;
  attributes: Record<string, string>;
  children: HtmlNode[];
}

/**
 * Reduce an element to the parts that carry meaning: its tag, its attributes
 * (sorted, so serialiser ordering cannot fail a comparison) and its children,
 * with whitespace-only text nodes dropped. Indentation therefore does not
 * matter, but a changed word does.
 */
const elementSignature = (el: Element): HtmlElementSignature => {
  const attributes: Record<string, string> = {};
  for (const name of el.getAttributeNames().toSorted()) {
    attributes[name] = el.getAttribute(name) ?? "";
  }

  const children: HtmlNode[] = [];
  for (const node of el.childNodes) {
    if (node.nodeType === node.TEXT_NODE) {
      const text = normalizeWhitespace(node.textContent);
      if (text) children.push(text);
    } else if (node.nodeType === node.ELEMENT_NODE) {
      children.push(elementSignature(node as Element));
    } else if (node.nodeType === node.COMMENT_NODE) {
      children.push({ comment: normalizeWhitespace(node.textContent) });
    }
  }

  return { tag: el.tagName.toLowerCase(), attributes, children };
};

const htmlSignature = (file: string): HtmlElementSignature => {
  const dom = new JSDOM(readText(file), { contentType: "text/html" });
  return elementSignature(dom.window.document.documentElement);
};

const compareHtml = (expected: string, actual: string): void => {
  expect(htmlSignature(actual), `HTML structure differs: ${actual}`).toEqual(
    htmlSignature(expected),
  );
};

type Comparator = (expected: string, actual: string) => void;

const COMPARATORS: Record<string, Comparator> = {
  ".json": compareJson,
  ".yaml": compareYaml,
  ".yml": compareYaml,
  ".html": compareHtml,
  ".htm": compareHtml,
};

const walkFiles = (dir: string, prefix = ""): string[] =>
  readdirSync(dir, { withFileTypes: true })
    .flatMap((entry) => {
      const rel = path.join(prefix, entry.name);
      if (entry.isDirectory()) return walkFiles(path.join(dir, entry.name), rel);
      return entry.isFile() ? [rel] : [];
    })
    .toSorted();

/**
 * Assert that every file under `expectedDir` exists under `actualDir` with
 * equivalent content. Extra files in `actualDir` are ignored — fixtures state
 * what must be true, not an exhaustive directory listing.
 */
export const compareTree = (expectedDir: string, actualDir: string): void => {
  for (const rel of walkFiles(expectedDir)) {
    const expected = path.join(expectedDir, rel);
    const actual = path.join(actualDir, rel);
    expect(existsSync(actual), `expected file is missing from the result tree: ${rel}`).toBe(true);
    const comparator = COMPARATORS[path.extname(rel).toLowerCase()] ?? compareText;
    comparator(expected, actual);
  }
};
